### PR TITLE
 codec/bytes: remove read trait from decode function of bytes

### PR DESCRIPTION
**Pull request by AndreMouche**
Hi,
This PR updates the decode function for `bytes`. It accesses slice directly instead of using read trait since read trait is not sufficient enough.
This PR could be reviewed after https://github.com/pingcap/tikv/pull/3060 merged.

### CODE REVIEW

File: src/util/codec/bytes.rs [hunk 1]
```diff
@@ -142,58 +142,56 @@ pub fn encoded_bytes_len(encoded: &[u8], desc: bool) -> usize {
...
-impl<'a> BytesDecoder for &'a [u8] {
-    fn remaining(&self) -> usize {
-        self.len()
-    }
-
-    fn peak_u8(&self) -> Option<u8> {
-        if self.is_empty() {
-            None
+pub fn decode_bytes(data: &mut BytesSlice, desc: bool) -> Result<Vec<u8>> {
+    let mut key = Vec::with_capacity(data.len());
```
Comment (BusyJay) [inline src/util/codec/bytes.rs]: Maybe can change to `data.len() / (ENC_GROUP_SIZE + 1) * ENC_GROUP_SIZE`.

**Comment by AndreMouche** [comment]
@BusyJay @hicqu @breeswish PTAL

File: src/coprocessor/codec/mysql/decimal.rs [hunk 2]
```diff
@@ -1890,7 +1889,12 @@ pub trait DecimalEncoder: Write {
 
 impl<T: Write> DecimalEncoder for T {}
 
-pub trait DecimalDecoder: BytesDecoder {
+pub trait DecimalDecoder: Read {
+    /// Get the remaining length in bytes of current reader.
+    fn remaining(&self) -> usize;
+
+    fn peak_u8(&self) -> Option<u8>;
```
Comment (Hoverbear) [inline src/coprocessor/codec/mysql/decimal.rs]: Please document this function.

**Comment by hicqu** [comment]
LGTM.

File: src/coprocessor/codec/mysql/decimal.rs (hunk 2 above)
Comment (BusyJay) [reply src/coprocessor/codec/mysql/decimal.rs]: @AndreMouche any updates?

File: src/util/codec/bytes.rs [hunk 3]
```diff
@@ -112,7 +112,7 @@ pub fn encoded_compact_len(mut encoded: &[u8]) -> usize {
     vn + (encoded.as_ptr() as usize - last_encoded)
 }
 
-pub trait CompactBytesDecoder: NumberDecoder {
+pub trait CompactBytesFromFileDecoder: NumberDecoder {
```
Comment (BusyJay) [inline src/util/codec/bytes.rs]: Why not remove it?

Comment (AndreMouche) [reply src/util/codec/bytes.rs]: no, since the function `apply_plain_cf_file ` cannot access a slice directly.

**Comment by AndreMouche** [comment]
@BusyJay @hicqu PTAL

**Comment by AndreMouche** [comment]
friendly ping @breeswish @BusyJay

File: src/util/codec/bytes.rs (hunk 3 above)
Comment (breeswish) [reply src/util/codec/bytes.rs]: Why this trait is renamed? From the interface of the trait, I don't see the reason that it is restricted to "from file".

**Comment by AndreMouche** [comment]
`CompactBytesFromFileDecoder ` means it only does decode for data from a file. We rename it in case of misusing it for byte slice for byte slice, it has better performance to use `decode_compact_bytes` than use `CompactBytesFromFileDecoder `.

**Pull request merged**